std/file.d

```diff
@@ -670,20 +651,214 @@ version(Windows) private void writeImpl(const(char)[] name, const(FSChar)* namez
+        else version(Windows)
+        {
+            DWORD numWritten = void;
+            WriteFile(fd, buffer.ptr + sum, cast(uint) cnt, &numWritten, null);
+        }
+        else
+            static assert(0, "Unsupported OS");
+
+        if (numWritten != cnt)
             break;
-        sum += numwritten;
+        sum += numWritten;
     }
-    cenforce(sum == size && CloseHandle(h), name, namez);
+
+    version(Posix)
+        cenforce(sum == size && core.sys.posix.unistd.close(fd) == 0, name, namez);
+    else version(Windows)
+        cenforce(sum == size && CloseHandle(fd), name, namez);
+    else
+        static assert(0, "Unsupported OS");
 }
 
+
+/++
+    Creates a file with a random name in $(LREF tempDir) and returns the name
+    of the file. If data is passed in, then the file is written with that data;
+    otherwise, it's empty.
+
+    The idea is that this creates a temporary file for testing or whatever
+    other purpose a temporary file might be needed for. However, it will only
+    be deleted if it's explicitly deleted or if the OS decides to delete it as
+    some OSes do with files in temp directories on startup or shutdown. So,
+    unlike $(REF tmpfile,std,stdio), the file has a name, and it can have all
+    of the things done to it that one might typically do with a file (including
+    reopening it after closing it).
+
+    The file is created with R/W permissions. On POSIX systems, the permissions
+    are restricted to the current user, though the effective permissions are
+    modified by the process' umask in the usual way.
+
+    Params:
+        prefix = Prefix to the random portion of the file name.
+        suffix = Suffix to the random portion of the file name (e.g. for a file
+                 extension).
+        buffer = Data to populate the file with.
+
+    Returns:
+        The name of the generated file.
+
+    Throws:
+        $(LREF FileException) if it fails to create the file.
+
+    See_Also:
+        $(REF tmpfile,std,stdio)
+  +/
+string createTempFile(const void[] buffer = null) @safe
+{
+    return createTempFile(null, null, buffer);
+}
+
+/// Ditto
+string createTempFile(const(char)[] prefix, const(char)[] suffix, const void[] buffer = null) @trusted
+{
+    import std.path : absolutePath, baseName, buildPath, dirName;
+    import std.random : isSeedable, Random, unpredictableSeed;
+
+    // We do this rather than use rndGen to avoid any problems caused by rndGen
+    // being reseeded or reused after being copied (which could easily lead to
+    // the program being able to figure out what the supposedly random name
+    // would be).
+    static Random random;
+    static bool initialized;
+    if (!initialized)
+    {
+        import std.algorithm.iteration : map;
+        import std.range : repeat;
+        static if (isSeedable!(Random, typeof(map!((a) => unpredictableSeed)(repeat(0)))))
+            random.seed(map!((a) => unpredictableSeed)(repeat(0)));
+        else
```

Comment:
This method of seeding calls `unpredictableSeed` 624 times to initialize the MersenneTwisterEngine's internal array. Currently this just means multiplying a number 624 times modulo 2^^31-1 and XOR-ing the result against the clock each time. This is pointless but fairly harmless, although it means that initially each number will have the same high bit. It would be faster and give a better result to just seed MersenneTwisterEngine with a single number. Also, if a proposal like https://github.com/dlang/phobos/pull/5230 to make `unpedictableSeed` more "unpredictable" is implemented, this mode of seeding could become pathological.

I hope this comment will be unnecessary due to the code having been changed either to use `rndGen` or to use a more lightweight PRNG than the Mersenne Twister.

Reply:
I just copied what std.random was doing for initializing `Random`. I do think that Cybershadow raised a legitimate concern with `rndGen`, though honestly I think that the bigger concern is the fact that `Random` is a value type, making it stupidly easy to reuse random numbers across the program, and the risk of it being reseeded isn't all that high and causing problems - particularly when this is all TLS, and worst case, it has to go through every random number used to generate a temporary file previously where that file hasn't been deleted (which isn't great, but it's not terribly likely either). So, while I do think that using `rndGen` is a concern, I'm not sure that it's so big a concern that we shouldn't do it - especially if it's so large (which certainly wasn't something that I thought of).